extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/tuple/sql/ArrayOfDoublesSketchSetBaseOperatorConversion.java

```diff
@@ -0,0 +1,136 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.query.aggregation.datasketches.tuple.sql;
+
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexLiteral;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.sql.SqlFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.SqlOperator;
+import org.apache.calcite.sql.type.OperandTypes;
+import org.apache.calcite.sql.type.SqlOperandCountRanges;
+import org.apache.calcite.sql.type.SqlTypeFamily;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.query.aggregation.PostAggregator;
+import org.apache.druid.query.aggregation.datasketches.tuple.ArrayOfDoublesSketchSetOpPostAggregator;
+import org.apache.druid.segment.column.RowSignature;
+import org.apache.druid.sql.calcite.expression.DruidExpression;
+import org.apache.druid.sql.calcite.expression.OperatorConversions;
+import org.apache.druid.sql.calcite.expression.PostAggregatorVisitor;
+import org.apache.druid.sql.calcite.expression.SqlOperatorConversion;
+import org.apache.druid.sql.calcite.planner.PlannerContext;
+
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+
+public abstract class ArrayOfDoublesSketchSetBaseOperatorConversion implements SqlOperatorConversion
+{
+  public ArrayOfDoublesSketchSetBaseOperatorConversion()
+  {
+  }
+
+  @Override
+  public SqlOperator calciteOperator()
+  {
+    return makeSqlFunction();
+  }
+
+  @Nullable
+  @Override
+  public DruidExpression toDruidExpression(
+      PlannerContext plannerContext,
+      RowSignature rowSignature,
+      RexNode rexNode
+  )
+  {
+    plannerContext.setPlanningError("%s can only be used on aggregates. " +
+        "It cannot be used directly on a column or on a scalar expression.", getFunctionName());
+    return null;
+  }
+
+  @Nullable
+  @Override
+  public PostAggregator toPostAggregator(
+      PlannerContext plannerContext,
+      RowSignature rowSignature,
+      RexNode rexNode,
+      PostAggregatorVisitor postAggregatorVisitor
+  )
+  {
+    final List<RexNode> operands = ((RexCall) rexNode).getOperands();
+    final List<PostAggregator> inputPostAggs = new ArrayList<>();
+    Integer nominalEntries = null;
+    Integer numberOfvalues = null;
+
+    for (int i = 0; i < operands.size(); i++) {
+      RexNode operand = operands.get(i);
+      if (i == 0 && operand.isA(SqlKind.LITERAL) && SqlTypeFamily.INTEGER.contains(operand.getType())) {
+        nominalEntries = RexLiteral.intValue(operand);
```

Comment:
These requirements could be better-plugged into the OperandTypeChecker, and if so, users would get better messages when using the function incorrectly. I'm willing to leave this for future work, though, if you aren't up to doing it in this patch.

Reply:
Hmm, not sure... I'm pretty new to the Calcite APIs.
Can you point me to a relevant example?

Reply:
Currently the OperandTypeChecker for this class is `OperandTypes.variadic(SqlOperandCountRanges.from(2))`. It would accept any call with at least two arguments. If the stuff in `OperandTypes` isn't enough to 

There's a couple of ways to do an OperandTypeChecker that has better checking:

1. You can combine various `OperandTypes` using `or`, `and`, `sequence`, etc.
2. You can make your own implementation of `SqlOperandTypeChecker`. For example, check out the `DoublesSketchListArgOperandTypeChecker` we have for `DS_GET_QUANTILES`.

Reply:
I had tried composing it with `or` and `sequence` but it didn't seem possible to support the required function signatures using them. (IIRC, `sequence` and `variadic` were not composable because `sequence` only can apply to `SqlSingleOperandTypeChecker` instances)

I will take a look at `DoublesSketchListArgOperandTypeChecker` as an example. However, note that even the existing Theta sketch SQL function doesn't seem to enforce it more strictly: https://github.com/apache/druid/blob/4493275d88d8a73ed1766ee07954d95c55dda862/extensions-core/datasketches/src/main/java/org/apache/druid/query/aggregation/datasketches/theta/sql/ThetaSketchSetBaseOperatorConversion.java#L120

Reply:
@gianm It isn't obvious to me exactly how to implement this part. Given that the parsing/validation added here is similar to what's already being done for other data sketch SQL functions, perhaps it's best to save your suggested improvements improvements for a future initiative.

Reply:
> Given that the parsing/validation added here is similar to what's already being done for other data sketch SQL functions, perhaps it's best to save your suggested improvements for a future initiative.

Yeah, I agree that the other sketch bindings could use better operand validation too. I'm OK with deferring this til the future.